### PR TITLE
file-drop gesture

### PR DESCRIPTION
**Pull request by AlexIljin**
Introducing a new ui gesture: `file-drop`!

When user drags-and-drops a file onto a `gadget`, the gadget may now react to the event by getting the list of file names and doing whatever it wants with them. Sample code is included in the listener. Currently the implementation only exists for Windows, but hopefully it's easy to support on other platforms.

The mechanism is different from the one used by `windows.dragdrop-listener` in that it doesn't rely on OLE, and therefore works on Vista+, even when you drag-drop files across apps with different security tokens. I.e. it works even when the OLE fails. Plus, making it a generic ui gesture, we give all platforms a chance to participate, whatever their underlying technologies might be.

The listener's reaction to the `file-drop` is also different from the one implemented by the `dragdrop-listener`. I decided to put the array of dropped file names onto the listener stack. I think it's more generic (and more useful) than simply running the dropped scripts: this way you may drop all kinds of files and handle them in various ways. Running them as scripts is just one of the possibilities (the one I have never used, frankly).

There is one issue: the listener must opt in to `accept-files`, and I'm not sure how to do it correctly. There is always the option of including the call in the user rc-file, but it feels like there should be a way to enable the feature by default on Windows.

Let me know what you think. I'm not sure about placing and naming of the `windows.dropfiles` vocab. Improvement suggestions are welcome!

### CODE REVIEW

**Comment by AlexIljin** [comment]
If someone really wants to immediately run a script dropped onto the listener window, there are at least two options:
- if only one file was dropped, check its extension and run it if it has the ".factor" suffix;
- run files as scripts if some modifier is depressed, such as A+ or C+.

**Comment by AlexIljin** [comment]
Come on, guys, this is an awesome feature!

**Comment by mrjbq7** [comment]
It's pretty cool.  Are you going to re-implement ``windows.dragdrop-listener`` in terms of your new ``filedrop`` gesture?  So that instead of adding to the stack, it tries to ``run-file`` them?

**Comment by AlexIljin** [comment]
The two drop catching mechanisms are independent. The vision is to have the `windows.dragdrop-listener` and the `windows.dropfiles` as the two top prongs of a "Y", both outputting the `file-drop` instances at the bottom.

My current plan is to modify `windows.dragdrop-listener` slightly, so that it generates a `file-drop` gesture instead of calling `run-file` directly. The listener then would handle the gesture however it feels like. By deferring the reaction to the listener, we are making the behavior platform-agnostic (as opposed to having a `windows.*` vocab tell it what to do).

I'm not entirely sure how useful it is to run scripts by dropping them onto the listener. I was never tempted to do that, and, judging by the number of years the vocab spent in `unmaintained`, I'm not the only one who feels that way. Right now I feel that the add-to-stack action would be simple, predictable and useful behavior.

If there are strong proponents of the drop-to-run behavior (maybe it's super-standard on MacOS, or something), then I can modify the listener's reaction to `file-drop` so that when a single file is dropped, and the file name has the ".factor" suffix, it will be executed via `run-file`. Otherwise an array of file names will be added to the listener's stack. Holding the `A+` modifier may then force a single dropped factor-file onto the stack as well.

I'm also looking into ways to combine the two mechanisms to get the benefits of both, if that's at all possible. If such combination is indeed possible, I'll have to add some additional gestures like `drag-over`, `drag-accept`, etc. Having that kind of gesture vocabulary will then open doors to native Factor object dragging, which I may or may not be able to implement.

**Comment by AlexIljin** [comment]
The `windows.dragdrop-listener` uses a universal COM API that can be used for pieces of text, images and anything, really, that can be copied into the system clipboard. Currently it only implements receiving of the `CF_HDROP` format. COM doesn't work across app security tokens.

The `windows.dropfiles` can only be used to accept the files (lists of file and folder names), but it works across security tokens (because the cross-process memory access needed to get to the file list is handled internally by the OS).

**Comment by catb0t** [comment]
any new developments on this? shame it's been open for so long, maybe the file conflicts can be merged manually @AlexIljin

**Comment by AlexIljin** [comment]
Let me have a look...

**Comment by AlexIljin** [comment]
I've resolved the conflicts, made a few improvements, and I think it's ready to be merged.

A question remains open, though: how do I make the listener call `accept-files` on startup?

I also made `windows.dragdrop-listener` produce the `file-drop` gestures, so both sources can be handled in the same manner by the gesture receiver, although I had to cheat a little bit. During the COM drag-and-drop operation Windows doesn't send the WM_MOUSEMOVE events, so the `hand-gadget` symbol doesn't get updated. I have added some code to update it, but it's not very sophisticated. As you can see, it simply sets `hand-gadget` to be the first child of the `world` that subscribed to the dragdrop. This works for the listener, but is not a generic solution. Some help on how to do this properly would be very much appreciated!

**Comment by AlexIljin** [comment]
Currently I didn't implement the option to "run the dropped file if it's a script", because we didn't decide on how it should be done (whether it should be controlled by the keyboard modifiers, and if so, which modifiers do what).

File: basis/windows/dropfiles/dropfiles-docs.factor [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    init-message-filter
+} ;
+
+ARTICLE: "about-dragdrop" "File drag-and-drop in Windows"
+"There are two mechanisms in Windows that can be used to drag-and-drop files across applications:"
+{ $list
+    { { $snippet "WM_DROPFILES" } " - was introduced back in the early days, it is a message that's posted to a window's message queue after the user has dropped some files on it. While handling the message, the application can fetch the list of the dropped files and the mouse position of the drop." }
+    { { $snippet "IDropTarget" } " - an OLE reinvention of the same. It provides more fine-grained capabilities of dynamically accepting or rejecting the drop based on the mouse location and the contents of the drop, while the user still drags the files over the window." }
+}
+"Windows Vista has introduced some security features that made it impossibe for the OLE to work between two applications with different security tokens. E.g. if one of the applications is ran with administrative privileges, and the other is without, the OLE drag-and-drop will not work between them."
```
Comment (mrjbq7) [inline basis/windows/dropfiles/dropfiles-docs.factor]: s/impossibe/impossible

**Comment by mrjbq7** [comment]
51 commits!  Now you have me curious.

File: basis/windows/dragdrop-listener/dragdrop-listener.factor [hunk 2]
```diff
@@ -1,26 +1,18 @@
...
-
-: filenames-from-hdrop ( hdrop -- filenames )
-    dup filecount-from-hdrop <iota>
-    [
-        2dup f 0 DragQueryFile 1 + ! get size of filename buffer
-        dup WCHAR <c-array>
-        [ swap DragQueryFile drop ] keep
-        utf16n alien>string
-    ] with map ;
+CONSTANT: E_OUTOFMEMORY -2147024882 ! 0x8007000e
```
[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: I'm not sure whether `0x8007000e` would be signed (I would expect it to be unsigned) and how that would compare to the negative error code (I would expect they would **not** be equal). I'll make a test run and change it if it works against my expectations.

File: basis/windows/dropfiles/dropfiles-docs.factor (hunk 1 above)
Comment (AlexIljin) [reply basis/windows/dropfiles/dropfiles-docs.factor]: Well spotted, thanks!

File: basis/windows/dropfiles/dropfiles.factor [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+: filenames-from-hdrop ( hdrop -- filenames )
+    dup filecount-from-hdrop <iota>
+    [
+        2dup f 0 DragQueryFile 1 + ! get size of filename buffer
+        dup WCHAR <c-array>
+        [ swap DragQueryFile drop ] keep
+        utf16n alien>string
+    ] with map ;
+
+! : point-from-hdrop ( hdrop -- loc )
```
[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dropfiles/dropfiles.factor]: This will be useful if we want to find out the exact mouse coordinates of the drop location within a `world`. Currently I rely on `hand-gadget` to be updated with the drop target gadget, but it is feasible to have gadgets that handle drops differently depending on the location. In any case, it presumes a future protocol extension, which isn't there at the moment. Maybe active mouse-with-drop tracking, visual feedback about actions to be performed on drop, etc.

In short, this is some that may be useful in the future.

File: basis/windows/dropfiles/dropfiles.factor [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+: handle-wm-dropfiles ( hdrop -- )
+    <alien> [ filenames-from-hdrop dropped-files set-global ] [ DragFinish ] bi
+    key-modifiers <file-drop> hand-gadget get-global propagate-gesture ;
+
+! The ChangeWindowMessageFilter has a global per-process effect, and so is the
+! list of wm-handlers. Therefore, there is no benefit in using the stricter
+! ChangeWindowMessageFilterEx approach. Plus, the latter is not in Vista.
+: (init-message-filter) ( -- )
+    "ChangeWindowMessageFilter" "user32" dlsym? [
```
[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dropfiles/dropfiles.factor]: The question really is - how do we handle HWNDs (window handles) that are saved in the image? Do we save them in any way, and if we do, do they become invalid and have to be recreated somehow?

The idea is you need to call `hwnd-accept-files` per window handle, and it calls the `init-message-filter`.

I wonder if the value of `\ init-message-state get-global` is saved in the image? I never thought about it...

[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dropfiles/dropfiles.factor]: Well, apparently it **is** saved. This means that I do need a startup hook at least to reset it to `f`.

File: basis/windows/dragdrop-listener/dragdrop-listener.factor (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: Nope. Those are different constants.
```
-2147024882 0x8007000e = .
f
```

File: basis/windows/dropfiles/dropfiles.factor (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dropfiles/dropfiles.factor]: OK, I've added a startup hook to do `f \ init-message-filter set-global`.

**Comment by AlexIljin** [comment]
@mrjbq7: yes, we need to call `accept-files` for every world that wants to accept file drops. Maybe later we can do in on a per-gadget basis, and maybe add a standard property or something. Your ideas are very much welcome! But as you said, baby steps.

File: basis/windows/dropfiles/dropfiles.factor [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      { $ WM_DROPFILES $ WM_COPYDATA $ WM_COPYGLOBALDATA }
+      [ MSGFLT_ADD ChangeWindowMessageFilter drop ] each
```
[the remark this answers is not in the document]
Comment (AlexIljin) [reply basis/windows/dropfiles/dropfiles.factor]: OK, done.

**Comment by AlexIljin** [comment]
OK, all comments are taken care of. Please review again and/or merge.

File: basis/windows/dragdrop-listener/dragdrop-listener.factor [hunk 6]
```diff
@@ -84,6 +73,10 @@ SYMBOL: +listener-dragdrop-wrapper+
 >>
 
 : dragdrop-listener-window ( -- )
-    world get handle>> hWnd>> dup <listener-dragdrop>
-    +listener-dragdrop-wrapper+ get-global com-wrap
-    [ RegisterDragDrop check-ole32-error ] with-com-interface ;
+    world get dup <listener-dragdrop>
+    +listener-dragdrop-wrapper+ get-global com-wrap [
+        [ handle>> hWnd>> ] dip
+        2dup RegisterDragDrop dup E_OUTOFMEMORY =
```
Comment (mrjbq7) [inline basis/windows/dragdrop-listener/dragdrop-listener.factor]: Why do you check ``E_OUTOFMEMORY`` here  and then call ``ole-initialize`` and call it again?

**Comment by mrjbq7** [comment]
Merged, thanks!

File: basis/windows/dragdrop-listener/dragdrop-listener.factor (hunk 6 above)
Comment (AlexIljin) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: Because that's exactly the error code it returns the first time if `ole-initialize` was never called.

`ole-initialize` needs to be called once per application run, before OLE-related functions start working. I didn't want to call it in a startup hook in case we never actually use `RegisterDragDrop`. And I couldn't assume nobody else called it before, so I need to call it at some point.

The way I did it here, it will be called only once, and the rest of the time the first `RegisterDragDrop` will succeed.

Theoretically, it is harmless to call it every time before each `RegisterDragDrop`, but it seemed a bit sloppy, since a one-time operation becomes all-the-time operation.

How do you usually do these things?

**Comment by AlexIljin** [comment]
Thank you for merging! : ) 👍

**Comment by AlexIljin** [comment]
@mrjbq7: "51 commits! Now you have me curious."
Well, it's mostly things we disagreed on over the years (removal of the unmaintained being one of them), my development environment specifics and some half-finished features like the `argon2` library FFI.
All the features I consider more or less complete are now officially in the main trunk.

Thanks for the active merging and reviewing these days, all of you guys! Catching up on my Factor is something I can't do any day, it needs tons of dedicated time, so it's awesome that you were there when I needed you.

I'm happy to be the sixth top [contributor](https://github.com/factor/factor/graphs/contributors) with - wow - 345 commits already : ))

File: basis/windows/dragdrop-listener/dragdrop-listener.factor (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonenst) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: The way to input your negative value with a literal hex is ```-0x7ff8fff2```.

Do we have something to input fixed width (32 or 64) literal signed (especially negative) numbers by their two-complement bits (like they do in the windows documentation) ? Something like this :
```factor
: >signed32 ( bits -- signed ) 0xffffffff bitxor 1 + neg ;
0x8007000E >signed32 . ! -2147024882
```
Maybe we can add a syntax in math.parser to have this natively ? I don't know, maybe reuse the L and LL suffix from C for 32bits and 64bits signed int ?

```factor
0xffffffffL ! -1
0xffffffffffffffffLL ! -1
```
 And add the corresponding printing functions ( .L .LL ?)
```
-1 .L !  0xffffffff
1 .L ! 0x00000001
```
What do you think ?

[the remark this answers is not in the document]
Comment (mrjbq7) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: We have ``math.bitwise:>signed``:

```factor
IN: scratchpad 0x8007000E 32 >signed .
-2147024882
```

Which I hadn't noticed until you asked that question.  I'm not sure about making it a parsing word when you can just do:

```factor
CONSTANT: foo $[ 0x8007000E 32 >signed ]
```

Which isn't as nice, but is as functional.

[the remark this answers is not in the document]
Comment (jonenst) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: Ha! I looked at math.bitwise but missed >signed. We should add it to http://docs.factorcode.org:8080/content/article-math.bitwise.html next to >even and >odd

[the remark this answers is not in the document]
Comment (mrjbq7) [reply basis/windows/dragdrop-listener/dragdrop-listener.factor]: Done.
